src/main/kotlin/org/elm/ide/refactoring/extractExpressionUI.kt

```diff
@@ -0,0 +1,40 @@
+package org.elm.ide.refactoring
+
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.util.Pass
+import com.intellij.refactoring.IntroduceTargetChooser
+import org.elm.lang.core.psi.ElmExpressionTag
+import org.elm.openapiext.isUnitTestMode
+import org.jetbrains.annotations.TestOnly
+
+fun showExpressionChooser(
+        editor: Editor,
+        exprs: List<ElmExpressionTag>,
+        callback: (ElmExpressionTag) -> Unit
+) {
+    if (isUnitTestMode) {
+        callback(MOCK!!.chooseTarget(exprs))
+    } else {
+        IntroduceTargetChooser.showChooser(editor, exprs, callback.asPass) { it.text }
+    }
+}
+
+private val <T> ((T) -> Unit).asPass: Pass<T>
+    get() = object : Pass<T>() {
+        override fun pass(t: T) = this@asPass(t)
+    }
+
+interface ExtractExpressionUi {
+    fun chooseTarget(exprs: List<ElmExpressionTag>): ElmExpressionTag
+}
+
+var MOCK: ExtractExpressionUi? = null
+@TestOnly
+fun withMockTargetExpressionChooser(mock: ExtractExpressionUi, f: () -> Unit) {
```

Comment:
nbd: You could potentially make the callsite simpler if you get rid of the `ExtractExpressionUi` interface and make this `mock: (List<ElmExpressionTag>) -> ElmExpressionTag`. 

Which would let you do

```kotlin
withMockTargetExpressionChooser(
        mock = { exprs ->
            //...
        }
) {
   // ...
}
```

Reply:
This interface will eventually grow to include additional hooks (choose anchor scope and the option to replace-all duplicate code expressions)